### PR TITLE
AnyEvent

### PR DESCRIPTION
**Pull request by andreyc21**
Hi. Can you please add support of AnyEvent backend for www::telegram::botapi?
I wrote the first version. Any suggestions are welcome.

### CODE REVIEW

**Comment by Robertof** [comment]
Hi,

thanks for your pull request! :heart: I really appreciate it.
Is there any specific need of using `AnyEvent::HTTP` explicitly? I'm asking because Mojolicious has transparent support for any kind of event loop, and even shows an example here on how to use AnyEvent: http://mojolicious.org/perldoc/Mojolicious/Guides/Cookbook#More-event-loops

See also [this](http://blog.kraih.com/post/43197898000/mojolicious-hack-of-the-day-anyevent).

Let me know, and thanks again!

**Comment by andreyc21** [comment]
Hi. I have some project already written on AnyEvent loop
I think it would be great if the module has flexibility and provide to developer the choice what backend to use. Like AnyEvent or JSON-MaybeXS do)

**Comment by Robertof** [comment]
Hello,

I think it could make sense to switch our currently hardcoded structure of two useragents to a dynamic one which allows expansion. I am currently thinking of a structure more or less like this:

```
WWW::Telegram::BotAPI::Agent
WWW::Telegram::BotAPI::Agent::LWP
WWW::Telegram::BotAPI::Agent::Mojo
WWW::Telegram::BotAPI::Agent::AnyEvent
```

Then, hot-loading of agents can be performed with LWP as the last choice and Mojo as the first.

However, I also prepared an example code for you which interlaces this module and `AnyEvent` with the `EV` backend. For your use case this could be useful.

```perl
#!/usr/bin/env perl
use strict;
use warnings;
use v5.10.1;
use EV; # enforce the usage of EV as the AnyEvent/Mojo::IOLoop backend
use AnyEvent;
use Mojo::UserAgent; # enforce the usage of Mojo::UserAgent
use WWW::Telegram::BotAPI;

# Instantiate the BotAPI.
my $api = WWW::Telegram::BotAPI->new (
    token => 'tok',
    async => 1
);

# Instantiate an AnyEvent condvar for demonstration purposes.
my $cv = AE::cv;

# Perform an asynchronous request with the standard BotAPI agent.
$api->getMe (sub {
    my (undef, $tx) = @_;
    warn ref $tx->error ? $tx->error->{message} : $tx->error unless $tx->success;
    $cv->send ($tx->res->json // {});
});

say "Awaiting end of asynchronous request run with AnyEvent...";
say "Bot nickname: ", $cv->recv->{result}{username} // "Something went wrong!";
```

What do you think?
Thanks!

**Comment by andreyc21** [comment]
Thanks for the great example) The only thing, I don't like to use  Mojolicious, only for Telegram notifications 

And
> WWW::Telegram::BotAPI::Agent
>WWW::Telegram::BotAPI::Agent::LWP
>WWW::Telegram::BotAPI::Agent::Mojo
>WWW::Telegram::BotAPI::Agent::AnyEvent

Looks great! I can try to write this if you do not mind)

**Comment by Robertof** [comment]
Hey,

thank you so much!
However, be aware that this should be a pretty big change, as it requires being backwards-compatible to the existing user-facing API and also will _probably_ break existing tests. If you're willing to write part of the code, I am more than happy to help you as well :+1: 

About the details of the implementation, I aim the `WWW::Telegram::BotAPI::Agent` API to be as extensive and module-indepent as possible. As such, we need a standardized interface for all the subclasses. This is my proposal:

- static method `supports_async` which returns a boolean value on whether this agent supports asynchronous requests.
- instance method `perform_request` which accepts the following hash as its parameters:
  ```perl
  $agent->perform_request (
    method => "POST", # or GET, or whatever.
    url => "https://xxx",
    data => { datagoes => 'here' }, # hash reference to POST data -- only acceptable if method is POST
                                    # this data is passed in the standard BotAPI format - which means that files are
                                    # specified using the Mojo::UserAgent syntax. See BotAPI.pm#133
                                    # the conversion between these kind of parameters and ones suitable for LWP
                                    # is already done in the existing code. It is now responsibility of the agent module
                                    # to do it, transparently.
    callback => sub {} # optional async callback to be called 
  )
  ```
  If no asynchronous callback is passed, then the request has to be executed blockingly and the response be returned immediately. In this case, the method returns a `WWW::Telegram::BotAPI::AgentResponse` object (see later for details).
  If an asynchronous callback is passed, then the method returns immediately. Whether the agent expects an event loop to be running, we need to discuss whether it should be executed by the agent or the user is expected to run it (if we'd like to keep the behavior as it is right now, we should not start any event loop.)
  To keep backwards compatibility, when using async mode the module should not perform any explicit error validation. As such, maybe we should give to the callback our response object. However, this is not acceptable when using `Mojo` as the user-agent because currently the callback receives the agent as the first argument and the transaction as the second one.

These two should be the main methods of the agent interface. About the response object, either we use a simple hash reference or we implement an object (let's call it `WWW::Telegram::BotAPI::AgentResponse` for now), for which I propose this implementation:
- gets instantiated with `WWW::Telegram::BotAPI::AgentResponse->new (error => "most meaningful error message or undef", response => { "json" => "object" })`
- has utility getters `success` (checks `!defined $self->{error}`), `error`, `response` and maybe others that I can't come up with right now.

Once that is defined, we also need to think about the module-loading mechanism. For that, I propose that the main module should try to load `WWW::Telegram::BotAPI::Agent::Mojo` first, and fallback to `WWW::Telegram::BotAPI::Agent::LWP` then (which is guaranteed to load -- since LWP is a dependency). I don't think that adding the `AnyEvent` agent to the default load list is a good choice. Maybe we could put it as the second choice after `Mojo` instead of directly fallbacking to LWP, as it has async support out of the box.

So, to recap, the steps to take if we want to proceed with this are:

- [ ] define the main `WWW::Telegram::BotAPI::Agent` superclass (with stub methods which just croak/die when called)
- [ ] define the main `WWW::Telegram::BotAPI::AgentResponse` class, if applicable
- [ ] implement agent modules
   - [ ] LWP
   - [ ] Mojo
   - [ ] AnyEvent
- [ ] switch the core module to the new API
- [ ] refactor existing tests (TBD)

So, as I said, if we want to go on with this and you want to help me, do it! And let me know what you think about. (it's late here so some things may be screwed up)

Thanks for your contribution!

**Comment by andreyc21** [comment]
Ok. I'll try)

**Comment by Robertof** [comment]
Okay, so let's work together! Do you have something you want to work on in particular? I can work on the rest ;)

P.S. Thanks!
